# 1.) Import the Credit Card Fraud Data From CCLE

In [1]:
import pandas as pd
from google.colab import drive
import matplotlib.pyplot as plt
import numpy as np

In [2]:
drive.mount('/content/gdrive/', force_remount = True)

Mounted at /content/gdrive/


In [3]:
df = pd.read_csv("/content/gdrive/MyDrive/Econ 441B/fraudTest.csv")

In [4]:
df.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2020-06-21 12:14:25  2291163933867244   
1           1   2020-06-21 12:14:33  3573030041201292   
2           2   2020-06-21 12:14:53  3598215285024754   
3           3   2020-06-21 12:15:15  3591919803438423   
4           4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2     Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3  Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4    Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   

   city_pop                     job         dob  \
0    333497     Mechanical engineer  1968-03-19   
1       302  Sales professional, IT  1990-01-17   
2     34496       Librarian, public  1970-10-21   
3     54767            Set designer  1987-07-25   
4      1126      Furniture designer  1955-07-06   

                          trans_num   unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

# 2.) Select four columns to use as features (one just be trans_date_trans)

In [5]:
df_select = df[["trans_date_trans_time", "category", "amt", "city_pop", "is_fraud"]]

In [6]:
df_select.columns

Index(['trans_date_trans_time', 'category', 'amt', 'city_pop', 'is_fraud'], dtype='object')

# 3.) Create a your own variable out of trans_date. Create dummies for factor vars

In [7]:
type(df_select["trans_date_trans_time"][0])

str

In [8]:
df_select["trans_date_trans_time"] = pd.to_datetime(df_select["trans_date_trans_time"])

<ipython-input-8-99f721e4ce0f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_select["trans_date_trans_time"] = pd.to_datetime(df_select["trans_date_trans_time"])


In [9]:
df_select["time_var"] = [i.weekday() for i in df_select["trans_date_trans_time"]]

<ipython-input-9-085e75a1d1b6>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_select["time_var"] = [i.weekday() for i in df_select["trans_date_trans_time"]]


In [10]:
X = pd.get_dummies(df_select, ["category"]).drop(["trans_date_trans_time", "is_fraud"], axis = 1)
y = df["is_fraud"]

In [11]:
X.head()

amt  city_pop  time_var  category_entertainment  category_food_dining  \
0   2.86    333497         6                       0                     0   
1  29.84       302         6                       0                     0   
2  41.28     34496         6                       0                     0   
3  60.05     54767         6                       0                     0   
4   3.19      1126         6                       0                     0   

   category_gas_transport  category_grocery_net  category_grocery_pos  \
0                       0                     0                     0   
1                       0                     0                     0   
2                       0                     0                     0   
3                       0                     0                     0   
4                       0                     0                     0   

   category_health_fitness  category_home  category_kids_pets  \
0                        0              0                   0   
1                        0              0                   0   
2                        1              0                   0   
3                        0              0                   0   
4                        0              0                   0   

   category_misc_net  category_misc_pos  category_personal_care  \
0                  0                  0                       1   
1                  0                  0                       1   
2                  0                  0                       0   
3                  0                  1                       0   
4                  0                  0                       0   

   category_shopping_net  category_shopping_pos  category_travel  
0                      0                      0                0  
1                      0                      0                0  
2                      0                      0                0  
3                      0                      0                0  
4                      0                      0                1

# XXX SKIP THIS WE WILL TALK ABOUT NEXT CLASS

In [12]:
resample_X = X
resample_y = y

# 5.) Train a Logistic regression.

In [13]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_normalized = scaler.fit_transform(resample_X)

In [14]:
from sklearn.linear_model import LogisticRegression

In [15]:
log_reg = LogisticRegression().fit(X_normalized, resample_y)

# 6.) The company you are working for wants to target at a False Positive rate of 5% what threshold should you use? (Use oversampled data)

In [16]:
# ASK chatgpt
# If you dont like the code ask in a different way


In [17]:
from sklearn.metrics import roc_curve
from scipy.interpolate import interp1d

# Use the predict_proba method to get the predicted probabilities
pred_probs = log_reg.predict_proba(X_normalized)[:, 1]

# Use the roc_curve function to get the false positive rate and true positive rate for different thresholds
fpr, tpr, thresholds = roc_curve(resample_y, pred_probs)

# Use the interp1d function to interpolate and get the threshold value that corresponds to a false positive rate of 5%
fpr_interp = interp1d(fpr, thresholds)
threshold = fpr_interp(0.05)

In [18]:
print(threshold)

0.010205897404389706


# 7.) If the company makes .02*amt on True transactions and loses -amt on False (Use original data)

In [19]:
df_temp = df_select.copy()

In [20]:
# Use the threshold from 6) to make binary predictions
df_temp["pred"] = (pred_probs >= threshold).astype(int)

In [21]:
df_temp = df_temp[["pred", "is_fraud", "amt"]]

In [22]:
df_temp.head()
# Which combinations of "pred" and "if_fraud" do we profit .02*amt, which cases do we lose -amt??

pred  is_fraud    amt
0     0         0   2.86
1     0         0  29.84
2     0         0  41.28
3     0         0  60.05
4     0         0   3.19

In [23]:
Earnings = 0
for i in range(0, len(df_temp)):
  if df_temp["pred"][i] == 0 and df_temp["is_fraud"][i] == 0:
    Earnings += df_temp["amt"][i] * 0.02
  elif df_temp["pred"][i] == 0 and df_temp["is_fraud"][i] == 1:
    Earnings -= df_temp["amt"][i]
Earnings

505444.4737999901

In [24]:
# Compare it with the default threshold = 0.5

In [25]:
df_temp["pred_default"] = log_reg.predict(resample_X)

/usr/local/lib/python3.8/dist-packages/sklearn/base.py:443: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(


In [26]:
Earnings_default = 0
for i in range(0, len(df_temp)):
  if df_temp["pred_default"][i] == 0 and df_temp["is_fraud"][i] == 0:
    Earnings_default += df_temp["amt"][i] * 0.02
  elif df_temp["pred_default"][i] == 0 and df_temp["is_fraud"][i] == 1:
    Earnings_default -= df_temp["amt"][i]
Earnings_default

-34637.56640000058

It can be seen that compared to a default threshold of 0.5, our strategy of implementing a threshold that keeps the false positive rate to 5% greatly improved the earnings from -34638 to 505444. Therefore, it is important to control the false negative rate as once the actual fraud is not detected, the company will lose all the money from that transaction. 

# 8.) Using Logistic Regression Lasso to inform you. Would you use the selected features in a trusted prediction model?

In [27]:
# Create an instance of the logistic regression model
log_reg = LogisticRegression(penalty='l1', C=0.1, solver='liblinear')

# Train the model using some training data
log_reg.fit(X_normalized, resample_y)

LogisticRegression(C=0.1, penalty='l1', solver='liblinear')

In [28]:
coef = log_reg.coef_
print(coef)

[[ 0.32521956 -0.11155464  0.14321603 -0.05398682 -0.06137889  0.10518372
   0.019083    0.43061445 -0.05411532 -0.09681637 -0.08104287  0.32557017
   0.          0.          0.41099227  0.10514898 -0.83656656]]


In [29]:
# If most or all your variables go to 0 => Your data is garbage
# The regularization will tell us if our model has significance
# This of using coefficient strength similar to r^2

Having 2 out of 17 coefficients equal to 0 means that those two features have been deemed to be less important by the model, and have been removed from the model. Our model is doing ok because most of the coefs are not 0. It suggests that those features might be important for predicting the response variable, and should be included in the model. I would probably try different feature subsets and compare the performance of the model using cross-validation. And then identify the best subset of features.